### ISSUE TITLE
Body-Pix in detail

### ISSUE THREAD

**Issue by ajaichemmanam** [tensorflow/tfjs#3040]
@oveddan @nsthorat @tylerzhu-github @tafsiri @dsmilkov @pyu10055 
Sorry for asking attention from the devs,

I have been trying to understand the bodypix model to port it to python. (Specifically Resnet Model)

I have ported the model to python. It produces  8 results
from these output tensors.
['resnet_v1_50/displacement_bwd_2/BiasAdd:0', 'resnet_v1_50/displacement_fwd_2/BiasAdd:0', 'float_heatmaps:0', 'float_long_offsets:0', 'float_short_offsets:0', 'float_part_heatmaps:0', 'float_segments:0', 'float_part_offsets:0']

**What I understood**

- The sigmoid of segment output is taken. And a threshold value is used. 
- If > segmentation threshold, we keep the value, else we put -1 as value in the partheatmaps.

- The sigmoid of partheatmaps is taken and the max() of it gives the position and confidence score. we use the same positions in partoffsets to find the offset at that positions.
Actual part position = position(x,y) of max() in partheatmap * Output Stride + partoffset(x,y)

- Use the heatmaps and short offsets to find the keypoints like in posenet

**Need Details**
- I don't understand where the long offsets is used. Short offset for keypoint position and part offset for part position
- I understood how multiperson decoding of posenet work (fast greedy approach from the personlab research paper). Can you help me with the multiperson decoding algorithm used for bodypix part segmentation
- Please correct me If I am wrong in any understandings

**Comment by koji333** [comment]
> I don't understand where the long offsets is used. Short offset for keypoint position and part offset for part position

The caption Fig.1. of the Personlab paper (https://arxiv.org/abs/1803.08225) seems to answer your question.
"... while the latter two ('Person segmentation mask' and 'Long-range offsets') are used, along with the human pose detections, are used by the **"Instance Segmentation Module"** in order to predict person instance segmentation masks."

>  Can you help me with the multiperson decoding algorithm used for bodypix part segmentation

The entry point (the **"Instance Segmentation Module"**) seems to be here:
(1) https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/body_pix_model.ts#L885
```
async segmentMultiPersonParts(
  input: BodyPixInput,
  config: MultiPersonInstanceInferenceConfig = MULTI_PERSON_INSTANCE_INFERENCE_CONFIG): Promise<PartSegmentation[]>
```
The neural network predicts by calling 'predictForMultiPersonInstanceSegmentationAndPart':
https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/body_pix_model.ts#L915
Post-processing calls 'decodePersonInstancePartMasks' function:
https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/body_pix_model.ts#L959

(2) https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/multi_person/decode_instance_masks.ts#L85
```
export async function decodePersonInstancePartMasks(
  segmentation: tf.Tensor2D, longOffsets: tf.Tensor3D,
  partSegmentation: tf.Tensor2D, poses: Pose[], height: number, width: number,
  stride: number, [inHeight, inWidth]: [number, number], padding: Padding,
  minPoseScore = 0.2, refineSteps = 8, minKeypointScore = 0.3,
  maxNumPeople = 10): Promise<PartSegmentation[]>
```

Then 'decodeMultiplePartMasksCPU' is called:
https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/multi_person/decode_instance_masks.ts#L117

(3)https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/multi_person/decode_multiple_masks_cpu.ts#L152
```
export function decodeMultiplePartMasksCPU(
  segmentation: Uint8Array, longOffsets: Float32Array,
  partSegmentaion: Uint8Array, posesAboveScore: Pose[], height: number,
  width: number, stride: number, [inHeight, inWidth]: [number, number],
  padding: Padding, refineSteps: number,
  numKptForMatching = 5): Int32Array[]
```

Then 'matchEmbeddingToInstance' function is called: 
https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/multi_person/decode_multiple_masks_cpu.ts#L173

(4) https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/multi_person/decode_multiple_masks_cpu.ts#L80
```
function matchEmbeddingToInstance(
  location: Pair, longOffsets: Float32Array, poses: Pose[],
  numKptForMatching: number, [padT, padL]: [number, number],
  [scaleX, scaleY]: [number, number], outputResolutionX: number,
  [height, width]: [number, number], stride: number,
  refineSteps: number): number
```

Then 'getEmbedding' function, where **"longOffsets"** seems to be used, is called:
https://github.com/tensorflow/tfjs-models/blob/2562fedcd156d4e5ef7d5fbe7067ef25fd2adc37/body-pix/src/multi_person/decode_multiple_masks_cpu.ts#L92

Though I shallowly read the code, I have somehow feeling of understanding the code with the explanation of this site:
https://medium.com/tensorflow/introducing-bodypix-real-time-person-segmentation-in-the-browser-with-tensorflow-js-f1948126c2a0
